clippy_lints/src/modulo_arithmetic.rs

```diff
@@ -0,0 +1,135 @@
+    /// ```
+    pub MODULO_ARITHMETIC,
+    restriction,
+    "any modulo arithmetic statement"
+}
+
+declare_lint_pass!(ModuloArithmetic => [MODULO_ARITHMETIC]);
+
+struct OperandInfo {
+    string_representation: Option<String>,
+    is_negative: bool,
+    is_integral: bool,
+}
+
+fn analyze_operand(operand: &Expr, cx: &LateContext<'_, '_>, expr: &Expr) -> Option<OperandInfo> {
+    match constant(cx, cx.tables, operand) {
+        Some((Constant::Int(v), _)) => match cx.tables.expr_ty(expr).kind {
+            ty::Int(ity) => {
+                let value = sext(cx.tcx, v, ity);
+                return Some(OperandInfo {
+                    string_representation: Some(value.to_string()),
+                    is_negative: value < 0,
+                    is_integral: true,
+                });
+            },
+            ty::Uint(_) => {
+                return Some(OperandInfo {
+                    string_representation: None,
+                    is_negative: false,
+                    is_integral: true,
+                });
+            },
+            _ => {},
+        },
+        Some((Constant::F32(f), _)) => {
+            return Some(floating_point_operand_info(&f));
+        },
+        Some((Constant::F64(f), _)) => {
+            return Some(floating_point_operand_info(&f));
+        },
+        _ => {},
+    }
+    None
+}
+
+fn floating_point_operand_info<T: Display + PartialOrd + From<f32>>(f: &T) -> OperandInfo {
+    OperandInfo {
+        string_representation: Some(format!("{:.3}", *f)),
+        is_negative: *f < 0.0.into(),
+        is_integral: false,
+    }
+}
+
+fn might_have_negative_value(t: &ty::TyS<'_>) -> bool {
+    t.is_signed() || t.is_floating_point()
+}
+
+impl<'a, 'tcx> LateLintPass<'a, 'tcx> for ModuloArithmetic {
+    fn check_expr(&mut self, cx: &LateContext<'a, 'tcx>, expr: &'tcx Expr) {
+        match &expr.kind {
+            ExprKind::Binary(op, lhs, rhs) | ExprKind::AssignOp(op, lhs, rhs) => {
+                if let BinOpKind::Rem = op.node {
+                    let lhs_operand = analyze_operand(lhs, cx, expr);
+                    let rhs_operand = analyze_operand(rhs, cx, expr);
+                    if_chain! {
+                        if let Some(lhs_operand) = lhs_operand;
+                        if let Some(rhs_operand) = rhs_operand;
+                        then {
+                            if lhs_operand.is_negative ^ rhs_operand.is_negative {
+                                span_note_and_lint(
+                                    cx,
+                                    MODULO_ARITHMETIC,
+                                    expr.span,
+                                    &format!(
+                                        "you are using modulo operator on constants with different signs: `{} % {}`",
+                                            lhs_operand.string_representation.unwrap(),
+                                            rhs_operand.string_representation.unwrap()
```

Comment:
NIT: 1 indentation level too much

Reply:
Reduced by extracting helper functions: `check_const_operands()` and `check_non_const_operands()`